testutil: ensure good treatment of argv on non-Unix platforms

levitte opened the pull request:
From a Unix point of view, some other platform families have certain
quirks.  Windows command prompt doesn't expand globs into actual file
names, so we must do this.  VMS has some oddity with argv pointer size
that can cause crashes if you're not careful (by copying it to a less
surprising pointer size array).

The fixups already exist and are used in the apps/ code.  However, the
testutil code started using the opt routines from apps/ without
including the non-Unix fixups.  This change fixes that.

levitte commented:
WIP until I've verified that the fix works everywhere

petrovr commented:
Richard Levitte wrote:
>  From a Unix point of view, some other platform families have certain
> quirks.  Windows command prompt doesn't expand globs into actual file
> names, so we must do this.  VMS has some oddity with argv pointer size
> that can cause crashes if you're not careful (by copying it to a less
> surprising pointer size array).
[SNIP]
So slowly I lost my mingw* knowledge, but mingw* binaries  support 
something similar.
On prima vista I cannot remember how to activate it.

Regards,
Roumen

levitte commented:
I finally got this to work right on OpenVMS at least.

levitte commented:
I'll deal with the appveyor failure after the weekend

levitte commented:
> I'll deal with the appveyor failure after the weekend

It was actually easier than I thought...

levitte committed:
VMS: move copy_argc to its own module and make it an aux source

copy_argv was never initialization code.

Make it self-cleaning too.

levitte committed:
testutil: ensure good treatment of argv on non-Unix platforms

From a Unix point of view, some other platform families have certain
quirks.  Windows command prompt doesn't expand globs into actual file
names, so we must do this.  VMS has some oddity with argv pointer size
that can cause crashes if you're not careful (by copying it to a less
surprising pointer size array).

The fixups already exist and are used in the apps/ code.  However, the
testutil code started using the opt routines from apps/ without
including the non-Unix fixups.  This change fixes that.

For VMS' sake, libtestutil gets an app_malloc() shim, to avoid sucking
in all of apps/apps.c.

levitte commented:
I started over, as it turned out that the VMS specific code was oddly organised.

levitte commented:
This finally builds and passes test [*] on VMS as well. Taken out of WIP.

-----

[*] apart from the shlib load test, but that's subject for a separate PR

paulidale commented on test/testutil/apps_mem.c:
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright 1995-2018 The OpenSSL Project Authors. All Rights Reserved.
```
??? shouldn't this just be 2019 ???

levitte replied:
copying and pasting, you know...  Fixing

levitte commented:
Merged.

ce506d27ab5e7d17dfe3fe649768a0d19b6c86ee testutil: ensure good treatment of argv on non-Unix platforms
9b542d72d2e7d4893a11b2e87628d9ac8637b954 VMS: move copy_argc to its own module and make it an aux source
